include/nbl/builtin/hlsl/workgroup/arithmetic.hlsl

```diff
@@ -15,64 +16,81 @@ namespace hlsl
+    T retVal = incl_wst(accessor);
+    accessor.main.workgroupExecutionAndMemoryBarrier();
+    return retVal;
+}
 
 template<typename T, class Binop, class SharedAccessor>
-struct exclusive_scan
+T exclusive_scan(T value, NBL_REF_ARG(SharedAccessor) accessor)
 {
-    struct inclusive_scan_t : subgroup::inclusive_scan<T, Binop> {}; // Yes, inclusive scan subgroup op is used for exclusive workgroup ops
-    T operator()(T value)
+    accessor.main.workgroupExecutionAndMemoryBarrier();
+    WSHT wsh = WSHT::create(Binop::identity(), _NBL_HLSL_WORKGROUP_SIZE_);
+    wsh(value, accessor);
+    WSTT(true) excl_wst = WSTT(true)::create(Binop::identity(), wsh);
+    T retVal = excl_wst(accessor);
+    accessor.main.workgroupExecutionAndMemoryBarrier();
+    return retVal;
+}
+
+#undef WSHT
+#undef WSTT
+
+/**
+ * Gives us the sum (reduction) of all ballots for the workgroup.
+ *
+ * Only the first few invocations are used for performing the sum 
+ * since we only have `uballotBitfieldCount` amount of uints that we need 
+ * to add together.
+ * 
+ * We add them all in the shared array index after the last DWORD 
+ * that is used for the ballots. For example, if we have 128 workgroup size,
+ * then the array index in which we accumulate the sum is `4` since 
+ * indexes 0..3 are used for ballots.
+ */ 
+template<class SharedAccessor>
+uint ballotBitCount(NBL_REF_ARG(SharedAccessor) accessor)
+{
+    accessor.main.set(uballotBitfieldCount, 0u);
+    accessor.main.workgroupExecutionAndMemoryBarrier();
+    if(gl_LocalInvocationIndex < uballotBitfieldCount)
     {
-        SharedAccessor accessor;
-        accessor.main.workgroupExecutionAndMemoryBarrier();
-        WorkgroupScanHead<T, inclusive_scan_t, SharedAccessor> wsh = WorkgroupScanHead<T, inclusive_scan_t, SharedAccessor>::create(true, Binop::identity(), _NBL_HLSL_WORKGROUP_SIZE_);
-        wsh(value);
-        WorkgroupScanTail<T, Binop, SharedAccessor> wst = WorkgroupScanTail<T, Binop, SharedAccessor>::create(true, Binop::identity(), wsh.firstLevelScan, wsh.lastInvocation, wsh.scanStoreIndex);
-        T retVal = wst();
-        accessor.main.workgroupExecutionAndMemoryBarrier();
-        return retVal;
+        const uint localBallot = accessor.main.get(gl_LocalInvocationIndex);
+        const uint localBallotBitCount = countbits(localBallot);
+        uint dummy;
+        accessor.main.atomicAdd(uballotBitfieldCount, localBallotBitCount, dummy);
     }
-};
+    accessor.main.workgroupExecutionAndMemoryBarrier();
+    return accessor.main.get(uballotBitfieldCount);
+}
 
 #define WSHT WorkgroupScanHead<uint, subgroup::inclusive_scan<uint, binops::add<uint> >, SharedAccessor>
-#define WSTT WorkgroupScanTail<uint, binops::add<uint>, SharedAccessor>
+#define WSTT WorkgroupScanTail<uint, binops::add<uint>, WSHT, SharedAccessor, true>
 template<class SharedAccessor>
-uint ballotScanBitCount(in bool exclusive)
+uint ballotScanBitCount(in bool exclusive, NBL_REF_ARG(SharedAccessor) accessor)
```

Comment:
take two accessors, one for the ballot state and another for the scan, so you don't have to take backups of the acessor state and have extra barriers

Also have you considered doing the `ballotBitCount` with a WSHT + Broadcast? Before I didn't do it that way, but only because the code was awfully non-modular and a pain to set up